Sources/NIOOpenSSL/ByteBufferBIO.swift

```diff
@@ -0,0 +1,289 @@
+internal func openSSLBIOGetsFunc(bio: UnsafeMutablePointer<BIO>?, buf: UnsafeMutablePointer<CChar>?, len: CInt) -> CInt {
+    return -2
+}
+
+/// The OpenSSL entry point for `BIO_ctrl`. We don't support most of these.
+internal func openSSLBIOCtrlFunc(bio: UnsafeMutablePointer<BIO>?, cmd: CInt, larg: CLong, parg: UnsafeMutableRawPointer?) -> CLong {
+    switch cmd {
+    case BIO_CTRL_GET_CLOSE:
+        return CLong(CNIOOpenSSL_BIO_get_shutdown(bio))
+    case BIO_CTRL_SET_CLOSE:
+        CNIOOpenSSL_BIO_set_shutdown(bio, CInt(larg))
+        return 1
+    case BIO_CTRL_FLUSH:
+        return 1
+    default:
+        return 0
+    }
+}
+
+internal func openSSLBIOCreateFunc(bio: UnsafeMutablePointer<BIO>?) -> CInt {
+    return 1
+}
+
+internal func openSSLBIODestroyFunc(bio: UnsafeMutablePointer<BIO>?) -> CInt {
+    return 1
+}
+
+
+/// An OpenSSL BIO object that wraps `ByteBuffer` objects.
+///
+/// OpenSSL extensively uses an abstraction called `BIO` to manage its input and output
+/// channels. For NIO we want a BIO that operates entirely in-memory, and it's tempting
+/// to assume that OpenSSL's `BIO_s_mem` is the best choice for that. However, ultimately
+/// `BIO_s_mem` is a flat memory buffer that we end up using as a staging between one
+/// `ByteBuffer` of plaintext and one of ciphertext. We'd like to cut out that middleman.
+///
+/// For this reason, we want to create an object that implements the `BIO` abstraction
+/// but which use `ByteBuffer`s to do so. This allows us to avoid unnecessary memory copies,
+/// which can be a really large win.
+final class ByteBufferBIO {
+    /// Pointer to the backing OpenSSL BIO object.
+    ///
+    /// Generally speaking OpenSSL wants to own the object initialization logic for a BIO.
+    /// This doesn't work for us, because we'd like to ensure that the `ByteBufferBIO` is
+    /// correctly initialized with all the state it needs. One of those bits of state is
+    /// a `ByteBuffer`, which OpenSSL cannot give us, so we need to build our `ByteBufferBIO`
+    /// *first* and then use that to drive `BIO` initialization.
+    ///
+    /// Because of this split initialization dance, we elect to initialize this data structure,
+    /// and have it own building an OpenSSL `BIO` structure.
+    private var bioPtr: UnsafeMutablePointer<BIO>
+
+    /// Owned pointer to the backing OpenSSL BIO object.
+    ///
+    /// This pointer is safe to use elsewhere, as it has increased the reference to the backing
+    /// `BIO`. This makes it safe to use with OpenSSL functions that require an owned reference
+    /// (that is, that consume a reference count).
+    ///
+    /// Note that the BIO may not remain useful for long periods of time: if the `ByteBufferBIO`
+    /// object that owns the BIO goes out of scope, the BIO will have its pointers invalidated
+    /// and will no longer be able to send/receive data.
+    internal var bio: UnsafeMutablePointer<BIO> {
```

Comment:
can we please make this a function called `ownedBIO` or `createBIO` or something? Not happy with either name suggestion but I'd like it to be a function so it's more clear it may have side-effects and also to stick more info there (ie. returns a +1 reference)